api.md

```diff
@@ -0,0 +1,242 @@
+# GOSwap Stats API
+
+TODO finish documenting descriptions
+TODO document all fields
+TODO make the actual API look like this (WARNING: it currently does not)
+
+### list tokens
+
+`/v1/tokens`
+
+`
+{
+  "tokens": [
+    {
+      "name": "string",
+      "symbol": "string",
+      "decimals": 123,
+      "address": "0xstring"
+    }
+  ]
+}
+`
+
+### get token
+
+`/v1/tokens/{address}`
+
+`
+{
+  "token": {
+    "name": "string",
+    "symbol": "string",
+    "decimals": 123,
+    "address": "0xstring"
+  }
+}
+`
+
+### list pairs
+
+`/v1/pairs`
+
+`
+{
+  "pairs": [
+    {
+      "index": 123,
+      "pair": "STRING-STRING",
+      "address": "0xstring",
+      "token0": "0xstring",
+      "token1": "0xstring"
+    }
+  ]
+}
+`
+
+### get pair
+
+`/v1/pairs/{address}`
+
+`
+{
+  "pair": {
+    "index": 123,
+    "pair": "STRING-STRING",
+    "address": "0xstring",
+    "token0": "0xstring",
+    "token1": "0xstring"
+  }
+}
+`
+
+### list stats totals
+
+Returns a sum of stat totals across all tokens/pairs that are `time_frame`
+apart, between `time_start` and `time_end`.
+
+TODO defaults for `time_` fields
+TODO do we want string encapsulated floats?
+
+```
+/v1/stats/totals
+?time_frame=1h
+?time_start=RFC3339-date
+?time_end=RFC3339-date
+```
+
+`
+{
+  "stats": [
+    {
+      "time":"RFC3339-date",
+      "volumeUSD": 123.123,
+      "liquidityUSD": 123.123
+    }
+  ]
+}
+`
+
+### get all token stats
+
+return token stats across all tokens between `time_start` and `time_end` that
+are `time_frame` apart.
+
+TODO should this only return a sum? (it's easy to query eg past 24h already)
+TODO defaults for `time_` fields
+TODO do we want string encapsulated floats?
+
+```
+/v1/stats/tokens
+?time_frame=1h
+?time_start=RFC3339-date
+?time_end=RFC3339-date
+```
+
+```
+{
+  "stats": [
```

Comment:
Should the results be separated by the tokens?  Or could remove this endpoint and only have the stats by token or pair. Or perhaps in your comment this only returns sums. But maybe this endpoint should be merged into /v1/tokens or /v1/pairs (sort of like it is now), just with the stats data as fields on the Pair/Token. It seems with this proposal you'd have to call /v1/tokens to get the token data, then /v1/stats/tokens to get the summed up data, then merge those yourself.

Reply:
yea I agree it's weird to have histograms across all tokens interspersed, thanks, will get rid of it. after thinking about it, think we just remove 'interval' and force sums for that endpoint between given start and end, so it's easy to do things like 7d or 24h. really wanted a 'cheap' endpoint just to list tokens/pairs, it's kinda standard api thing I guess, I know primary use case atm is just to get addresses to use but still - I follow on merging the two, it's a little kludgy I guess but if you really just want to get a name/address for a token it's a lot of TMI (as API stands, no way to get name / symbol of token without having handle on its address other than grabbing it from list endpoint - could change that too). in a way the 'list all' is kind of just an optimization for looking up a token details by its address to fill in (N requests), to my eye anyway. 

I think this endpoint still may need a little sugar to specify a subset of tokens possibly but it can wait (eg `?in=ETH,WGO,YFI` for just getting one page of a list or something along those lines, can wait for use case tho).

yeah, flow wise I was thinking we need a few things really for a token or pair:

* list all
* get volume sums of 24h, 7d, etc
* get latest liquidity (currently merged into ^)
* get latest price (token)
* get fee sums of 24h, 7d, etc
* get histogram of volume / price

so tried to logically divide this up in the same manner. I think each column being a request (staring at uniswap page here) is kinda what I'm going for, or on token / pair detail page need histograms on top of latest numbers but a request for each is still pretty good I think. if you take this a step further you get `/stats/tokens/{address}/volume` and things like that, to allow more specificity basically. but, makes my head spin a bit thinking about caching all of this. anyway, sums!

Reply:
What if we add a param to include stats on /v1/pairs and /v1/tokens to include sums/liquidity? Like ?stats=24h would return stats, without it it would just return the static info.

Reply:
yea, that could work, makes me want to rethink api tho (ie why have `/stats` at all then?). also pretty odd in the backend itself, have to create a new struct and merge the two and return, so gains seem kind of minimal in that sense, just picking to do it in the backend instead of the frontend at that point I think. kinda per comment, hoping to map each endpoint to 1 db call to start out, not least because it makes it super easy to cache, main benefit is just simplicity. guessing at some point we end up doing more complicated stats reports but not sure how that will bubble up either, api is flexible to that extent. only concern i'm tinkering with having the sums separate is whether to have `map[address]stat{}` in the json but it's also kind of just a convenience, in my head the front end will be something like this for the 'sums' page:

* get /tokens -> add to map[address]token (same idea for pairs), probably tote this around for various pages
* get token sums/liquidity/price -> sort by favorite column (eg volume), do quick lookup in tokens map from ^ to get token symbol to display instead of the address when rendering each row, et voila?

this reminds me I need to average price for the sums endpoint, or could use latest like liquidity. I guess either one is interesting, it's possible now to get buckets and get average for anyone interested, so nvm (punt).

Reply:
oh nvm we have pair and symbol in the respective pair and token buckets now, so the previous comment is moot, originally only pair buckets had pair which I was gonna bring up but this seems uniform now and I think solves the rift here? I think having them in the bucket works well enough, at worst it wastes a few (very few) bytes but it's not so big a deal. if we wanted to show full description (maybe for hover I guess?) it's probably only gonna be on token/pair detail page I think so that gets cheaper. anyway, proceeding with sums.